doc: Add a repo-level README

anubhav756 opened the pull request:
This PR includes adding a top-level README for the repo.

The README focusses on an overall guidance of using the subpackages included in this mono-repo.

> [!NOTE]
> This README would be simplified, especially the code snippets, once we align the APIs of the `toolbox-core` and orchestration-specific SDKs, like `toolbox-langchain`, in a future PR.

anubhav756 commented:
/gcbrun

twishabansal commented on README.md:
```diff
@@ -1,342 +1,361 @@
-applications, enabling advanced orchestration and interaction with GenAI models.
+[![License: Apache
+2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+[![PyPI - Python
```
We're supporting python 3.13 as well. This shows 3.9 to 3.12 only.

anubhav756 replied:
Added python 3.13 support in the metadata.

twishabansal commented on README.md:
```diff
@@ -1,342 +1,361 @@
+    formats (e.g., with parsed docstrings for LLM use), you might need to wrap
+    the loaded tools (e.g., using LangChain's
+    `StructuredTool.from_function(tool, parse_docstring=True)` as shown in the
+    `toolbox-core` README).
```
```suggestion
    `toolbox-core` README). Therefore, it is preferred to use the Orchestration framework specific SDKs when available. See all [available packages](https://github.com/googleapis/mcp-toolbox-sdk-python/blob/main/README.md#available-packages)
```

twishabansal commented on README.md:
````diff
@@ -1,342 +1,361 @@
-```
+If you encounter issues or have questions, please check the existing [GitHub
+Issues](https://github.com/googleapis/genai-toolbox/issues) for the main Toolbox
+project. If your issue is specific to one of the SDKs and not found, feel free
````
```suggestion
project. If your issue is specific to one of the SDKs, please look for existing issues [here](https://github.com/googleapis/mcp-toolbox-sdk-python/issues) or 
```

twishabansal commented on README.md:
```diff
@@ -1,342 +1,361 @@
+If you encounter issues or have questions, please check the existing [GitHub
+Issues](https://github.com/googleapis/genai-toolbox/issues) for the main Toolbox
+project. If your issue is specific to one of the SDKs and not found, feel free
+to open a new issue in that repository.
```
```suggestion
open a new issue in that repository.
```

twishabansal reviewed (commented):
We need to make sure our interfaces in `toolbox-core` and `toolbox-langchain` align.

anubhav756 commented:
> We need to make sure our interfaces in `toolbox-core` and `toolbox-langchain` align.

As discussed offline, we will be doing the breaking changes in an upcoming PR.

twishabansal commented on README.md:
````diff
@@ -1,342 +1,346 @@
-```bash
-pip install toolbox-langchain
-```
+## Overview
````
nit: it shows 2 TOCs

anubhav756 replied:
Oops...the plugin I used seems to act a bit funny though. Thanks for pointing that out.

twishabansal commented on README.md:
```diff
@@ -1,342 +1,346 @@
-
-builder.add_node("agent", call_model)
-builder.add_node("tools", tool_node)
+# toolbox-core
```
Let's move the code under the correct bullet points?

anubhav756 replied:
Good idea. Thanks!

twishabansal commented on README.md:
```diff
@@ -1,342 +1,346 @@
+Issues](https://github.com/googleapis/genai-toolbox/issues) for the main Toolbox
+project. If your issue is specific to one of the SDKs, please look for existing
+issues [here](https://github.com/googleapis/mcp-toolbox-sdk-python/issues) or
+open a new issue in that repository.
```
```suggestion
open a new issue in this repository.
```

twishabansal commented on README.md:
```diff
@@ -1,342 +1,346 @@
 
-> [!IMPORTANT]
-> You don't need to modify tool configurations to bind parameter values.
+| Feature                 | `toolbox-core`                                                                                                   | `toolbox-langchain`                                                                        |
```
I think it will be hard to update this table in the future with any other orchestration sdks. I would prefer if each row corresponds to a package and a column corresponds to a feature so that it's easier to extend. What do you think?

anubhav756 replied:
That's a good idea. Thanks!

twishabansal commented on README.md:
```diff
@@ -1,342 +1,346 @@
+    - [License](#license)
+    - [Support](#support)
+
+<!-- /TOC -->These SDKs act as clients for that service, abstracting away the
```
nit: please correcting formatting. see [this](https://screenshot.googleplex.com/AhttT4S3fs3RMdq)

twishabansal commented on README.md:
```diff
@@ -1,342 +1,346 @@
 
-graph.invoke({"messages": [HumanMessage(content="Do something with the tools")]})
+# Load all tools from all toolsets
+all_tools = sync_client.load_toolset()
```
This does not work yet. Should we add a note?

twishabansal reviewed (changes requested):
Thinking about the main README, I'm wondering if the number of `toolbox-langchain` code snippets is becoming a maintenance burden. To avoid having to update code examples in multiple places for every small change, perhaps we could reduce the snippets in the main README and link to the toolbox-langchain README for more detailed examples? What are your thoughts?

anubhav756 commented:
> Thinking about the main README, I'm wondering if the number of `toolbox-langchain` code snippets is becoming a maintenance burden. To avoid having to update code examples in multiple places for every small change, perhaps we could reduce the snippets in the main README and link to the toolbox-langchain README for more detailed examples? What are your thoughts?

I've kept them here for now, anticipating the upcoming API convergence between the Core and Orch SDKs (see [doc](https://docs.google.com/document/d/1kaBfCllqKJpIfAkissPNtwA8bjxT0eaLtmI27fzxxr8/edit?usp=sharing&resourcekey=0-0YEMYcEwMm9bpi0b-Tlt5A)). This convergence should allow us to easily showcase consistent snippets across both of these subpackages (along with the upcoming subpackage). Given this expected future state of reduced maintenance due to API alignment, do you think it's best to keep them here temporarily?

twishabansal commented:
> > Thinking about the main README, I'm wondering if the number of `toolbox-langchain` code snippets is becoming a maintenance burden. To avoid having to update code examples in multiple places for every small change, perhaps we could reduce the snippets in the main README and link to the toolbox-langchain README for more detailed examples? What are your thoughts?
> 
> I've kept them here for now, anticipating the upcoming API convergence between the Core and Orch SDKs (see [doc](https://docs.google.com/document/d/1kaBfCllqKJpIfAkissPNtwA8bjxT0eaLtmI27fzxxr8/edit?usp=sharing&resourcekey=0-0YEMYcEwMm9bpi0b-Tlt5A)). This convergence should allow us to easily showcase consistent snippets across both of these subpackages (along with the upcoming subpackage). Given this expected future state of reduced maintenance due to API alignment, do you think it's best to keep them here temporarily?

We could remove these once the APIs are aligned. Sounds good. Thanks!

kurtisvg reviewed (changes requested):
Should we replicate the contents of the other two READMEs here?

Or should e focus which packages are in these repo, and let the top level README try to answer questions like "which package should I use in which circumstances"?

kurtisvg commented:
I want us to optimize for the intersection of clarity and maintainability.

My current thought process is that we really need to communicate 2 things to users:
1. which package they should use. 
2. how to use the package

It's obviously more

For the first thing, I think the top-level README has to address it. Hopefully if they are at the package README they already known, but the package level READMEs should have a little INFO box that says something like "This is the <> package, intended to be used with <>. If you aren't sure which package to use, see the project README.". 

For the second thing, we have a common functionality (of across all packages), and package specific functionality. I think we _have_ to include the "how-to" for package-level functionality. Additionally, the READMEs are also present on the pypi package downloads. Given that, I think we _should_ include all the "how-to" (for both common and specific) in the package level READMEs because it's more clear.

However, I lean towards avoid putting the common functionality in the top-level README. It's both a second place to keep up to date, and it's less clear because a user may not read in it's entirely and miss the link to the package specific readme.

TL;DR: My 2c is that the intersection of clarity/maintainability is that top level README should cover "which package" and the package level README should cover "how to use".

anubhav756 commented:
> I want us to optimize for the intersection of clarity and maintainability.
> 
> My current thought process is that we really need to communicate 2 things to users:
> 
> 1. which package they should use.
> 2. how to use the package
> 
> It's obviously more
> 
> For the first thing, I think the top-level README has to address it. Hopefully if they are at the package README they already known, but the package level READMEs should have a little INFO box that says something like "This is the <> package, intended to be used with <>. If you aren't sure which package to use, see the project README.".
> 
> For the second thing, we have a common functionality (of across all packages), and package specific functionality. I think we _have_ to include the "how-to" for package-level functionality. Additionally, the READMEs are also present on the pypi package downloads. Given that, I think we _should_ include all the "how-to" (for both common and specific) in the package level READMEs because it's more clear.
> 
> However, I lean towards avoid putting the common functionality in the top-level README. It's both a second place to keep up to date, and it's less clear because a user may not read in it's entirely and miss the link to the package specific readme.
> 
> TL;DR: My 2c is that the intersection of clarity/maintainability is that top level README should cover "which package" and the package level README should cover "how to use".

Totally agree with your reasoning – makes way more sense to have everything needed for a package *in* its README for clarity. Even if we repeat some common setup info across the package READMEs, the potential gains from trying to maintain a single, separate repo-level 'how-to' seem pretty small compared to the confusion it could cause for users.

Just updated the main repo README:

* It's now focused just on helping users figure out *which package* they need (core vs orch-specific).
* Removed the detailed *how-to* sections (like core concepts, etc.) to keep it clean. Those details belong in the package READMEs now.
* Added a specific *"Which Package Should I Use?"* section upfront.
* Simplified the "Getting Started" > "Use the SDK" part - it just points to the "Available Packages" list for the specific README links now, so we don't have to repeat them (scales better with more packages! 👍).
